### ISSUE TITLE
CentOS http Status 404

### ISSUE THREAD

**Issue by GasparYYC** [OraOpenSource/OXAR#200]
Installed OXAR (downloaded today)
CentOS 7.2 (from Digital Ocean)
APEX 5.0.4
Oracle XE 11.2.0.1
ORDS 3.0.6
SQLCL 4.2.0

After installation http://_IP_/ shows:
404 Not Found

I noticed the following in the console:

`\* Installing ORDS *
###### ################################################################## 100.0%

Oct 24, 2016 9:25:40 PM oracle.dbtools.cmdline.ModifyConfigDir execute
INFO: Set config.dir to /etc in: /root/oxar/tmp/ords/ords.war
Oct 24, 2016 9:37:58 PM oracle.dbtools.rt.config.setup.SchemaSetup getInstallOrUpgrade
WARNING: Failed to connect to user ORDS_PUBLIC_USER jdbc:oracle:thin:@localhost:1521:xe
IO Error: Connection reset
Oct 24, 2016 9:37:58 PM oracle.dbtools.common.config.file.ConfigurationFilesBase update
INFO: Updated configurations: defaults, apex, apex_pu, apex_al, apex_rt
Oct 24, 2016 9:38:02 PM oracle.dbtools.rt.config.setup.SetProperty execute
INFO: Modified: /etc/ords/defaults.xml, settting: security.verifySSL = false
Oct 24, 2016 9:38:05 PM oracle.dbtools.rt.config.setup.SetProperty execute
INFO: Modified: /etc/ords/defaults.xml, settting: security.requestValidationFunction = wwv_flow_epg_include_modules.authorize
Oct 24, 2016 9:38:08 PM oracle.dbtools.rt.config.setup.SetProperty execute
INFO: Modified: /etc/ords/defaults.xml, settting: apex.excel2collection = true
Oct 24, 2016 9:38:12 PM oracle.dbtools.rt.config.setup.SetProperty execute
INFO: Modified: /etc/ords/defaults.xml, settting: apex.excel2collection.useSheetName = true

`

The only changes to the config.properties:
Used Dropbox to host the files - No errors accessing the files.
The OOS_APEX_ADMIN_EMAIL from admin to my email.

Any ideas??

**Comment by GasparYYC** [comment]
Whoa!
From the server reboot (after installation) until it magically worked (I didn't do anything, just kept trying) took ~45 minutes

I can see the APEX login window...

I'll retry the installation and report back..

**Comment by tschf** [comment]
Might be similar to an issue @martindsouza reported to me. Can you do the output of: `systemctl status tomcat@oxar`

**Comment by GasparYYC** [comment]
Looks like it started:

]# systemctl status tomcat@oxar
● tomcat@oxar.service - Apache Tomcat Web Application Container
   Loaded: loaded (/usr/lib/systemd/system/tomcat@oxar.service; enabled; vendor preset: disabled)
   Active: active (running) since Mon 2016-10-24 23:11:27 CDT; 14min ago
 Main PID: 5063 (java)
   CGroup: /system.slice/system-tomcat.slice/tomcat@oxar.service
           └─5063 /usr/bin/java -classpath /usr/share/tomcat/bin/bootstrap.jar:/usr/share/tomcat/bin/tomcat-juli.jar:/usr/share/java/commons-daem...

Oct 24 23:24:13 dcs2 server[5063]: Oct 24, 2016 11:24:13 PM org.apache.catalina.startup.HostConfig deployDirectory
Oct 24 23:24:13 dcs2 server[5063]: INFO: Deploying web application directory /var/lib/tomcat/webapps/i
Oct 24 23:24:13 dcs2 server[5063]: Oct 24, 2016 11:24:13 PM org.apache.catalina.startup.HostConfig deployDirectory
Oct 24 23:24:13 dcs2 server[5063]: INFO: Deployment of web application directory /var/lib/tomcat/webapps/i has finished in 389 ms
Oct 24 23:24:13 dcs2 server[5063]: Oct 24, 2016 11:24:13 PM org.apache.coyote.AbstractProtocol start
Oct 24 23:24:13 dcs2 server[5063]: INFO: Starting ProtocolHandler ["http-bio-8080"]
Oct 24 23:24:13 dcs2 server[5063]: Oct 24, 2016 11:24:13 PM org.apache.coyote.AbstractProtocol start
Oct 24 23:24:13 dcs2 server[5063]: INFO: Starting ProtocolHandler ["ajp-bio-8009"]
Oct 24 23:24:13 dcs2 server[5063]: Oct 24, 2016 11:24:13 PM org.apache.catalina.startup.Catalina start
Oct 24 23:24:13 dcs2 server[5063]: INFO: Server startup in 762729 ms

**Comment by GasparYYC** [comment]
Seems like APEX is responding after ~20 minutes after the Installation reboot.
Before it was showing a 404 and it now shows APEX Logon Screen.

**Comment by GasparYYC** [comment]
Tried again after restarting tomcat;
It took ~15 minutes to respond:
tail /usr/share/tomcat/logs/catalina.2016-10-24.log
.
.
.
Oct 24, 2016 11:26:04 PM oracle.ucp.common.UniversalConnectionPoolBase initInactiveConnectionTimeoutTimer
INFO: inactive connection timeout timer scheduled
Oct 24, 2016 11:41:24 PM oracle.dbtools.common.config.db.DatabasePoolFactory listConfiguration
INFO: Configuration properties for: |apex||
apex.excel2collection=true
apex.excel2collection.useSheetName=true
.
.
.

**Comment by GasparYYC** [comment]
Confirmed..
Takes about 30 minutes from the moment the machine is restarted until APEX Login.
Seems like about ~15 mins. then something shows in the log then after another ~15 mins. finally the logon screen opens..

Tomorrow I'll try with OpenJDK to see if there is any difference. Currently I am using Java JDK 1.1.8 (whatever is the latest today).

**Comment by martindsouza** [comment]
I think it's all dealing with the same issue (see #194) with tomcat taking a long time to get APEX up and running. @tschf is looking into it. No idea as to why this is happening.

**Comment by janihur** [comment]
A blind shot: not enough entropy http://stackoverflow.com/q/26431922/272735 ?

**Comment by tschf** [comment]
And how about with switching to the tomcat service I described earlier?
(away from tomcat@oxar). Any difference?

**Comment by GasparYYC** [comment]
@tschf I have followed the commands in order
After reboot I'm looking at the Catalina log file,

...
Oct 25, 2016 5:29:30 PM org.apache.catalina.core.StandardEngine startInternal
INFO: Starting Servlet Engine: Apache Tomcat/7.0.54
Oct 25, 2016 5:29:30 PM org.apache.catalina.startup.HostConfig deployWAR
INFO: Deploying web application archive /var/lib/tomcat/webapps/ords.war
Oct 25, 2016 5:29:50 PM oracle.dbtools.common.config.db.DatabasePools validatePool
INFO: Validating pool: |apex||
**<<This is when 10 mins. after tomcat continued loading>>**
Oct 25, 2016 5:39:48 PM oracle.dbtools.common.config.db.DatabasePools isValid
SEVERE: The connection pool named: apex is not correctly configured, due to the following error(s): IO Error: Connection reset
oracle.dbtools.common.jdbc.ConnectionPoolConfigurationException: The connection pool named: apex is not correctly configured, due to the following error(s): IO Error: Connection reset
        at oracle.dbtools.common.config.db.DatabasePools.validateConnection(DatabasePools.java:427)
        at oracle.dbtools.common.config.db.DatabasePools.access$100(DatabasePools.java:61)
        at oracle.dbtools.common.config.db.DatabasePools$1.call(DatabasePools.java:306)
        at oracle.dbtools.common.config.db.DatabasePools$1.call(DatabasePools.java:297)
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4742)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4739)
...


Now when I access the URL I get the ORDS web page (the one with the ORDS Logo and the pink band showing the error) with a 404 Not Found

**Comment by GasparYYC** [comment]
Success at 5:56 (CST)
...
Oct 25, 2016 5:39:54 PM oracle.dbtools.common.logging.LegacyLoggingAdaptor severe
SEVERE: ClientAbortException:  java.net.SocketException: Broken pipe (Write failed)
InternalServerException [statusCode=500, reasons=[An unexpected error with the following message occurred: null]]
        at oracle.dbtools.http.errors.ServletResponseExceptionMapper.mapError(ServletResponseExceptionMapper.java:84)
        at oracle.dbtools.http.errors.ErrorPageRenderer.<init>(ErrorPageRenderer.java:43)
        at oracle.dbtools.http.errors.ErrorPageRenderer.<init>(ErrorPageRenderer.java:35)
        at oracle.dbtools.rt.web.HttpEndpointBase.service(HttpEndpointBase.java:94)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:102)
        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1040)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:607)
        at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:314)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:745)
Caused by: ClientAbortException:  java.net.SocketException: Broken pipe (Write failed)
        at org.apache.catalina.connector.OutputBuffer.doFlush(OutputBuffer.java:371)
        at org.apache.catalina.connector.OutputBuffer.flush(OutputBuffer.java:333)
        at org.apache.catalina.connector.CoyoteOutputStream.flush(CoyoteOutputStream.java:101)
        at oracle.dbtools.http.servlet.NormalizedServletResponse$NormalizedOutputStream.flush(NormalizedServletResponse.java:918)
        at oracle.dbtools.http.servlet.NormalizedServletResponse.flushBuffer(NormalizedServletResponse.java:220)
        at oracle.dbtools.http.errors.ErrorPageRenderer.service(ErrorPageRenderer.java:97)
        at oracle.dbtools.http.errors.ErrorPageFilter.doFilter(ErrorPageFilter.java:109)
        at oracle.dbtools.http.filters.HttpFilter.doFilter(HttpFilter.java:46)
        at oracle.dbtools.http.filters.FilterChainImpl.doFilter(FilterChainImpl.java:64)
        at oracle.dbtools.http.auth.ForceAuthFilter.doFilter(ForceAuthFilter.java:44)
        at oracle.dbtools.http.filters.HttpFilter.doFilter(HttpFilter.java:46)
        at oracle.dbtools.http.filters.FilterChainImpl.doFilter(FilterChainImpl.java:64)
        at oracle.dbtools.http.filters.Filters.filter(Filters.java:47)
        at oracle.dbtools.http.entrypoint.EntryPoint.service(EntryPoint.java:83)
        at oracle.dbtools.http.entrypoint.EntryPointServlet.service(EntryPointServlet.java:47)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
        at oracle.dbtools.rt.web.HttpEndpointBase.dispatchableServices(HttpEndpointBase.java:123)
        at oracle.dbtools.rt.web.HttpEndpointBase.service(HttpEndpointBase.java:85)
        ... 21 more
Caused by: java.net.SocketException: Broken pipe (Write failed)
        at java.net.SocketOutputStream.socketWrite0(Native Method)
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
        at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
        at org.apache.coyote.http11.InternalOutputBuffer.realWriteBytes(InternalOutputBuffer.java:215)
        at org.apache.tomcat.util.buf.ByteChunk.flushBuffer(ByteChunk.java:480)
        at org.apache.coyote.http11.InternalOutputBuffer.flush(InternalOutputBuffer.java:119)
        at org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:799)
        at org.apache.coyote.Response.action(Response.java:174)
        at org.apache.catalina.connector.OutputBuffer.doFlush(OutputBuffer.java:366)
        ... 38 more

Oct 25, 2016 5:56:03 PM oracle.dbtools.common.config.db.DatabasePoolFactory listConfiguration
INFO: Configuration properties for: |apex||
apex.excel2collection=true
apex.excel2collection.useSheetName=true
cache.caching=false
cache.directory=/tmp/apex/cache
cache.duration=days
cache.expiration=7
cache.maxEntries=500
cache.monitorInterval=60
cache.procedureNameList=
cache.type=lru
db.hostname=localhost
db.port=1521
db.sid=xe
debug.debugger=false
debug.printDebugToScreen=false
error.keepErrorMessages=true
error.maxEntries=50
jdbc.DriverType=thin
jdbc.InactivityTimeout=1800
jdbc.InitialLimit=10
jdbc.MaxConnectionReuseCount=1000
jdbc.MaxLimit=60
jdbc.MaxStatementsLimit=10
jdbc.MinLimit=10
jdbc.statementTimeout=900
log.logging=false
log.maxEntries=50
misc.compress=
misc.defaultPage=apex
security.crypto.enc.password=******
security.crypto.mac.password=******
security.disableDefaultExclusionList=false
security.maxEntries=2000
security.requestValidationFunction=wwv_flow_epg_include_modules.authorize
security.verifySSL=false
db.password=******
db.username=APEX_PUBLIC_USER

Oct 25, 2016 5:56:03 PM oracle.ucp.common.UniversalConnectionPoolBase initInactiveConnectionTimeoutTimer
INFO: inactive connection timeout timer scheduled


I see the APEX logon screen.

**Comment by GasparYYC** [comment]
I changed the JDK from Oracle to OpenJDK (via alternatives command) and retried...
The server after about ~45 mins. did not even show the 404 error page..
I changed it back, restarted and APEX is up and running after about  ~36 mins.
Basically same behaviour as before...

**Comment by GasparYYC** [comment]
Do you know how to enable extended login in tomcat? Maybe I can try that way...
May provide more information...

**Comment by martindsouza** [comment]
@GasparYYC @tschf have either of you tried @janihur idea about entropy? (scroll up a bit for link). There's a few ideas in the SO answers.

**Comment by GasparYYC** [comment]
I did some more reading and Digital Ocean has a [How-To](https://www.digitalocean.com/community/tutorials/how-to-install-apache-tomcat-7-on-ubuntu-14-04-via-apt-get) that shows to use Install Tomcat.

Following the article's recommendations I edited the file: /usr/share/tomcat/conf/tomcat.conf
`JAVA_OPTS="-Djava.security.egd=file:/dev/./urandom -Djava.awt.headless=true -Xmx400m"`

And voila!
Now APEX shows the Logon Screen in about ~53 seconds from a cold server startup!
I've never seen APEX boot up so fast!

Note: I'm still using Oracle JDK.

I think this is done, big thanks to @janihur for pointing us in the right direction!

**Comment by GasparYYC** [comment]
Going a bit deeper.. 
  At least in my system the enthropy pool runs empty:
`# cat /proc/sys/kernel/random/entropy_avail
0
`

And according to  [DO-Article](https://www.digitalocean.com/community/tutorials/how-to-setup-additional-entropy-for-cloud-servers-using-haveged) this should not go < 1000

So I installed haveged:
`# yum install haveged
`
and make it to start at boot:
`# chkconfig haveged on`

I checked and the pool is now > 1500 consistently (it fluctuates)
I removed the custom configuration on /usr/share/tomcat/conf/tomcat.conf
Restarted the server..

and voila!
APEX login screen shows in under a minute..

**In Summary:**
  Based on my observations, there is no need to customize Tomcat config file.
  The solution is solely based on installing haveged package and set it to auto-start

**Comment by fuzziebrain** [comment]
The haveged workaround seems to work. Thanks @GasparYYC and @janihur.

**Comment by Nunalves** [comment]
This happened to me also, haven't tried the haveged, but i figured that going to the tomcat console and clicking the reload button on ords makes apex boot up, hope it gives an hint on what might be happening

**Comment by martindsouza** [comment]
@GasparYYC Thanks for all your detective work! We applied the patch and it's working fine now on cloud servers.
